core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -0,0 +1,424 @@
+	return &logBuffer{
+		lggr:          lggr.Named("KeepersRegistry.LogEventBufferV1"),
+		opts:          newLogBufferOptions(lookback, blockRate, logLimit),
+		lastBlockSeen: new(atomic.Int64),
+		queues:        make(map[string]*upkeepLogQueue),
+	}
+}
+
+// Enqueue adds logs to the buffer and might also drop logs if the limit for the
+// given upkeep was exceeded. It will create a new buffer if it does not exist.
+// Returns the number of logs that were added and number of logs that were  dropped.
+func (b *logBuffer) Enqueue(uid *big.Int, logs ...logpoller.Log) (int, int) {
+	buf, ok := b.getUpkeepQueue(uid)
+	if !ok || buf == nil {
+		buf = newUpkeepLogBuffer(b.lggr, uid, b.opts)
+		b.setUpkeepQueue(uid, buf)
+	}
+	latestBlock := latestBlockNumber(logs...)
+	if b.lastBlockSeen.Load() < latestBlock {
+		b.lastBlockSeen.Store(latestBlock)
+	}
+	blockThreshold := b.lastBlockSeen.Load() - int64(b.opts.lookback.Load())
+	if blockThreshold <= 0 {
+		blockThreshold = 1
+	}
+	return buf.enqueue(blockThreshold, logs...)
+}
+
+// Dequeue greedly pulls logs from the buffers.
+// Returns logs and the number of remaining logs in the buffer.
+func (b *logBuffer) Dequeue(block int64, blockRate, upkeepLimit, maxResults int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	start, end := getBlockWindow(block, blockRate)
+	return b.dequeue(start, end, upkeepLimit, maxResults, upkeepSelector)
+}
+
+// dequeue pulls logs from the buffers, depends the given selector (upkeepSelector),
+// in block range [start,end] with minimum number of results per upkeep (upkeepLimit)
+// and the maximum number of results (capacity).
+// Returns logs and the number of remaining logs in the buffer for the given range and selector.
+// NOTE: this method is not thread safe and should be called within a lock.
+func (b *logBuffer) dequeue(start, end int64, upkeepLimit, capacity int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	var result []BufferedLog
+	var remainingLogs int
+	for _, q := range b.queues {
+		if !upkeepSelector(q.id) {
+			// if the upkeep is not selected, skip it
+			continue
+		}
+		logsInRange := q.sizeOfRange(start, end)
+		if logsInRange == 0 {
+			// if there are no logs in the range, skip the upkeep
+			continue
+		}
+		if capacity == 0 {
+			// if there is no more capacity for results, just count the remaining logs
+			remainingLogs += logsInRange
+			continue
+		}
+		if upkeepLimit > capacity {
+			// adjust limit if it is higher than the actual capacity
+			upkeepLimit = capacity
+		}
+		logs, remaining := q.dequeue(start, end, upkeepLimit)
+		for _, l := range logs {
+			result = append(result, BufferedLog{ID: q.id, Log: l})
+			capacity--
+		}
+		remainingLogs += remaining
+	}
+	return result, remainingLogs
+}
+
+func (b *logBuffer) SetConfig(lookback, blockRate, logLimit uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.opts.override(lookback, blockRate, logLimit)
```

Comment:
Yes, to avoid extra maintenance we use a single object with atomic values that can be shared by multiple objects cross goroutines. The upkeep queue reads these values on demand and adjust capacity accordingly in `enqueue()` > `clean()`.